File: modules/data/src/main/java/org/gjt/jclasslib/structures/constants/ConstantReference.java
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.structures.constants;

import org.gjt.jclasslib.structures.CPInfo;
import org.gjt.jclasslib.structures.InvalidByteCodeException;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
    Base class for constant pool data structures which reference class members.

    @author <a href="mailto:jclasslib@ej-technologies.com">Ingo Kegel</a>
*/
public abstract class ConstantReference extends CPInfo {

    /** Length of the constant pool data structure in bytes. */
    public static final int SIZE = 4;

    /** <tt>class_index</tt> field. */
    protected int classIndex;
    /** <tt>name_and_type_index</tt> field. */
    protected int nameAndTypeIndex;
    
    public String getVerbose() throws InvalidByteCodeException {

        ConstantNameAndTypeInfo nameAndType = getNameAndTypeInfo();

        return classFile.getConstantPoolEntryName(classIndex) + "." +
               classFile.getConstantPoolEntryName(nameAndType.getNameIndex());
    }

    /**
        Get the index of the constant pool entry containing the
        <tt>CONSTANT_Class_info</tt> of this entry.
        @return the index
     */
    public int getClassIndex() {
        return classIndex;
    }
    
    /**
        Set the index of the constant pool entry containing the
        <tt>CONSTANT_Class_info</tt> of this entry.
        @param classIndex the index
     */
    public void setClassIndex(int classIndex) {
        this.classIndex = classIndex;
    }
    
    /**
        Get the index of the constant pool entry containing the
         <tt>CONSTANT_NameAndType_info</tt> of this entry.
        @return the index
     */
    public int getNameAndTypeIndex() {
        return nameAndTypeIndex;
    }

    /**
        Set the index of the constant pool entry containing the
         <tt>CONSTANT_NameAndType_info</tt> of this entry.
        @param nameAndTypeIndex the index
     */
    public void setNameAndTypeIndex(int nameAndTypeIndex) {
        this.nameAndTypeIndex = nameAndTypeIndex;
    }

    /**
        Get the class info for this reference.
        @return the class info.
        @throws InvalidByteCodeException
     */
    public ConstantClassInfo getClassInfo() throws InvalidByteCodeException {
        return (ConstantClassInfo)getClassFile().getConstantPoolEntry(classIndex, ConstantClassInfo.class);
    }

    /**
        Get the name and type info for this reference.
        @return the name and type info.
        @throws InvalidByteCodeException
     */
    public ConstantNameAndTypeInfo getNameAndTypeInfo() throws InvalidByteCodeException {
        return (ConstantNameAndTypeInfo)classFile.getConstantPoolEntry(
                    nameAndTypeIndex,
                    ConstantNameAndTypeInfo.class);
    }

    public void read(DataInput in)
        throws InvalidByteCodeException, IOException {
            
        classIndex = in.readUnsignedShort();
        nameAndTypeIndex = in.readUnsignedShort();
    }
    
    public void write(DataOutput out)
        throws InvalidByteCodeException, IOException {
        
        out.writeShort(classIndex);
        out.writeShort(nameAndTypeIndex);
    }
    
    public boolean equals(Object object) {
        if (!(object instanceof ConstantReference)) {
            return false;
        }
        ConstantReference constantReference = (ConstantReference)object;
        return super.equals(object) &&
               constantReference.classIndex == classIndex &&
               constantReference.nameAndTypeIndex == nameAndTypeIndex;
    }

    public int hashCode() {
        return super.hashCode() ^ classIndex ^ nameAndTypeIndex;
    }

}

File: modules/data/src/main/kotlin/org/gjt/jclasslib/structures/constants/ConstantReference.kt
/*
    This library is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public
    License as published by the Free Software Foundation; either
    version 2 of the license, or (at your option) any later version.
*/

package org.gjt.jclasslib.structures.constants

import org.gjt.jclasslib.structures.AbstractConstant
import org.gjt.jclasslib.structures.ClassFile
import org.gjt.jclasslib.structures.InvalidByteCodeException
import java.io.DataInput
import java.io.DataOutput

/**
 * Base class for constant pool data structures which reference class members.
 */
abstract class ConstantReference(classFile: ClassFile) : AbstractConstant(classFile) {

    /**
     * Index of the constant pool entry containing the
     * CONSTANT_Class_info of this entry.
     */
    var classIndex: Int = 0

    /**
     * Index of the constant pool entry containing the
     * CONSTANT_NameAndType_info of this entry.
     */
    var nameAndTypeIndex: Int = 0

    override val verbose: String
        @Throws(InvalidByteCodeException::class)
        get() = classFile.getConstantPoolEntryName(classIndex) + "." + classFile.getConstantPoolEntryName(nameAndTypeInfo.nameIndex)

    /**
     * Class info for this reference.
     */
    val classInfo: ConstantClassInfo
        @Throws(InvalidByteCodeException::class)
        get() = classFile.getConstantPoolEntry(classIndex, ConstantClassInfo::class.java)

    /**
     * Name and type info for this reference.
     * @throws InvalidByteCodeException
     */
    val nameAndTypeInfo: ConstantNameAndTypeInfo
        @Throws(InvalidByteCodeException::class)
        get() = classFile.getConstantPoolEntry(nameAndTypeIndex, ConstantNameAndTypeInfo::class.java)

    override fun readData(input: DataInput) {
        classIndex = input.readUnsignedShort()
        nameAndTypeIndex = input.readUnsignedShort()
    }

    override fun writeData(output: DataOutput) {
        output.writeShort(classIndex)
        output.writeShort(nameAndTypeIndex)
    }

    override val debugInfo: String
        get() = "with classIndex $classIndex and nameAndTypeIndex $nameAndTypeIndex"

    override fun equals(other: Any?): Boolean {
        if (other !is ConstantReference) {
            return false
        }
        return super.equals(other) && other.classIndex == classIndex && other.nameAndTypeIndex == nameAndTypeIndex
    }

    override fun hashCode(): Int = super.hashCode() xor classIndex xor nameAndTypeIndex

}
